src/ucp/dt/datatype_iter.inl

```diff
@@ -280,42 +281,62 @@ ucp_datatype_iter_next_ptr(const ucp_datatype_iter_t *dt_iter,
+    } else {
+        ucs_assertv(memh_index < ucs_popcount(dt_reg->md_map),
+                    "memh_index=%d md_map=0x%" PRIx64, memh_index,
+                    dt_reg->md_map);
+        return dt_reg->memh[memh_index];
+    }
+}
+
 /*
  * Returns a pointer to next chunk of data as IOV entry of registered memory
  * (could be done only on some datatype classes)
  *
  * @param memh_index  Index of UCT memory handle (within the memory domain map
  *                    which was passed to @ref ucp_datatype_iter_mem_reg, or
  *                    UCP_NULL_RESOURCE to pass UCT_MEM_HANDLE_NULL.
+ *
+ * @return Number of iov elements.
  */
-static UCS_F_ALWAYS_INLINE void
+static UCS_F_ALWAYS_INLINE size_t
 ucp_datatype_iter_next_iov(const ucp_datatype_iter_t *dt_iter,
-                           ucp_rsc_index_t memh_index, size_t max_length,
-                           ucp_datatype_iter_t *next_iter, uct_iov_t *iov)
+                           size_t max_length, ucp_rsc_index_t memh_index,
+                           unsigned dt_mask, ucp_datatype_iter_t *next_iter,
+                           uct_iov_t *iov, size_t max_iov)
 {
-    /* TODO support IOV datatype */
-    ucs_assert(dt_iter->dt_class == UCP_DATATYPE_CONTIG);
-
-    if (memh_index == UCP_NULL_RESOURCE) {
-        iov[0].memh = UCT_MEM_HANDLE_NULL;
+    ucs_assert(max_iov >= 1);
+    if (ucp_datatype_iter_is_class(dt_iter, UCP_DATATYPE_CONTIG, dt_mask)) {
+        iov[0].length = ucp_datatype_iter_next_ptr(dt_iter, max_length,
+                                                   next_iter, &iov[0].buffer);
+        iov[0].memh   = ucp_datatype_iter_uct_memh(&dt_iter->type.contig.reg,
+                                                   memh_index);
+        iov[0].stride = 0;
+        iov[0].count  = 1;
+        return 1;
+    } else if (ucp_datatype_iter_is_class(dt_iter, UCP_DATATYPE_IOV, dt_mask)) {
+        return ucp_datatype_iter_iov_next_iov(dt_iter, max_length, memh_index,
+                                              next_iter, iov, max_iov);
     } else {
-        ucs_assert(memh_index < ucs_popcount(dt_iter->type.contig.reg.md_map));
-        iov[0].memh = dt_iter->type.contig.reg.memh[memh_index];
+        next_iter->offset = dt_iter->offset; /* Silence compiler warning */
+        ucs_bug("unsupported datatype %s",
+                ucp_datatype_class_names[dt_iter->dt_class]);
+        return 0;
     }
-
-    iov[0].length   = ucp_datatype_iter_next_ptr(dt_iter, max_length, next_iter,
-                                                 &iov[0].buffer);
-    iov[0].stride   = 0;
-    iov[0].count    = 1;
 }
 
 /*
  * Copy iterator position
  */
-static UCS_F_ALWAYS_INLINE
-void ucp_datatype_iter_copy_from_next(ucp_datatype_iter_t *dt_iter,
-                                      const ucp_datatype_iter_t *next_iter,
-                                      unsigned dt_mask)
+static UCS_F_ALWAYS_INLINE void
+ucp_datatype_iter_copy_from_next(ucp_datatype_iter_t *dt_iter,
```

Comment:
does `next_iter` must be actually "next"? maybe generatize to "set/reset" and document that iters must be from the same datatype object?

Reply:
yes, it must be coming from iterator function. it's only copying the position of the iterator to avoid extra overhead. Maybe rename it to ucp_datatype_iter_copy_position()?